rasa/core/lock_store.py

```diff
@@ -0,0 +1,227 @@
+import asyncio
+import json
+import logging
+import typing
+from typing import Text, Optional, Dict, Union
+
+from async_generator import asynccontextmanager, async_generator, yield_
+
+from rasa.core.lock import TicketLock
+
+if typing.TYPE_CHECKING:
+    pass
+
+logger = logging.getLogger(__name__)
+
+ACCEPTED_LOCK_STORES = ["in_memory", "redis"]
+
+
+class LockError(Exception):
+    pass
+
+
+class LockStore(object):
+    def __init__(self, lifetime: int = 60) -> None:
+        self.lifetime = lifetime
+
+    @staticmethod
+    def find_lock_store(store=None) -> "LockStore":
+        if store is None or store.type is None or store.type == "in_memory":
+            lock_store = InMemoryLockStore()
+        elif store.type == "redis":
+            lock_store = RedisLockStore(host=store.url, **store.kwargs)
+        else:
+            raise ValueError(
+                "Cannot create `LockStore` of type '{}'. One of the following "
+                "`LockStore` types need to be specified: {}."
+                "".format(store.type, ", ".join(ACCEPTED_LOCK_STORES))
+            )
+
+        logger.debug(
+            "Connected to lock store {}.".format(lock_store.__class__.__name__)
+        )
+
+        return lock_store
+
+    def create_lock(self, conversation_id: Text) -> TicketLock:
+        """Create and save a new `TicketLock` for `conversation_id`."""
+
+        lock = TicketLock(conversation_id)
+        self.save_lock(lock)
+        return lock
+
+    def get_lock(self, conversation_id: Text) -> Optional[TicketLock]:
+        """Fetch lock for `conversation_id` from storage."""
+
+        raise NotImplementedError
+
+    def delete_lock(self, conversation_id: Text) -> None:
+        """Delete lock for `conversation_id` from storage."""
+
+        raise NotImplementedError
+
+    def save_lock(self, lock: TicketLock) -> None:
+        """Commit `lock` to storage."""
+
+        raise NotImplementedError
+
+    def issue_ticket(
+        self, conversation_id: Text, lifetime: Union[float, int] = None
+    ) -> int:
+        """Issue new ticket for lock associated with `conversation_id`.
+
+        Creates a new lock if none is found.
+        """
+
+        lock = self.get_or_create_lock(conversation_id)
+
+        lifetime = lifetime if lifetime is not None else self.lifetime
+        ticket = lock.issue_ticket(lifetime)
+
+        self.save_lock(lock)
+        return ticket
+
+    @asynccontextmanager
+    @async_generator
+    async def lock(
+        self, conversation_id: Text, attempts: int = 60, wait: Union[int, float] = 1
```

Comment:
```suggestion
        self, conversation_id: Text, attempts: int = 60, timeout_in_seconds: Union[int, float] = 1
```

or `wait_time_in_seconds`. However, I think it makes sense to include the unit in the name.